Feature #392 tcdiag_line_type

JohnHalleyGotway opened the pull request:
This is a large set of changes to the functionality and output of `tc_pairs`. This also include updates to `tc_stat` to handle the `tc_pairs` changes. This work is sufficiently mature to be included in the beta4 release since it adds a new TCDIAG line type, and we need to give the METplus Analysis tools time to test with that during the beta5 development cycle.

I expect that additional modifications will be required in the beta5 cycle, such as:
- Profiling `tc_stat` processing TCDIAG output from `tc_pairs` to see why its relatively slow.
- Enhance `tc_pairs` to handle LSDIAG_DEV data.
- Refine the logic for how different diagnostics data sources are integrated with existing track data.
- Add logic to handling some LSDIAG diagnostics currently being ignored.

## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[Yes]**</br>
If **yes**, please describe:</br>
  - Adds `-diag` command line option for `tc_pairs`.
  - Adds `diag_name` entry to the TC-Pairs config file.
  - Adds `diag_convert_map` entry to the TC-Pairs config file.
  - Adds `diag_thresh_name` and `diag_thresh_val` TC-Stat config file entries and corresponding `-diag_thresh` job command option.
  - Adds `init_diag_thresh_name` and `init_diag_thresh_val` TC-Stat config file entries and corresponding `-init_diag_thresh` job command option.

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes]**</br>
If **yes**, please describe:</br>
  - Adds the TCDIAG line type to the output of TC-Pairs. If diagnostics data is provided and added to a TC track, TC-Pairs writes a TCDIAG line immediately after the TCMPR line to which it corresponds.

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
  - Tested manually many times using TCDIAG and LSDIAG_RT input datasets.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
  - @georgemccabe please review the config file changes and consider what impacts they have to the TCPairs METplus wrapper. Please confirm that the only GHA diffs are based on the newly added tests in unit_tc_pairs.xml and unit_tc_stat.xml.
  - @jvigh please review the User's Guide changes to TC-Pairs and note the [Storm Diagnostics](https://met--2315.org.readthedocs.build/en/2315/Users_Guide/tc-pairs.html#storm-diagnostics) section where your input is needed during the beta5 cycle. Please review the config file changes, command line options, and overall design considerations.
  - @KathrynNewman please test this out in seneca:/d1/projects/MET/MET_pull_requests/met-11.0.0/met-11.0.0-beta4/feature_392/MET-feature_392_tcdiag_line_type. Make sure it works as expected with good inputs. Make sure that it fails as expected with bad input. Please let me know if the wording of any error messages should be tweaked. Make sure to test both `tc_pairs` and `tc_stat`, especially the `-diag_thresh` and `-init_diag_thresh` job command options.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**
  - See the chapters for `tc_pairs` and `tc_stat`.

- [x] Do these changes include sufficient testing updates? **[Yes]**
  - New tests for `tc_pairs` and `tc_stat`. I've already updated the input data sets with the necessary input data files.

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
  - New output from `tc_pairs` and `tc_stat` with DIAGNOSTICS in the name.

- [x] Please complete this pull request review by **[Thurs 10/27/22]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Development** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

JohnHalleyGotway committed:
Per #392, add hooks for storing diagnostic info alongside the track info.

JohnHalleyGotway committed:
Per #392, add hooks for writing a new TCDIAG output line type.

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_392_tcdiag_line_type

JohnHalleyGotway committed:
Per #392, continue working on hooks to store/write TCDIAG output lines interleaved with the TCMPR data.

JohnHalleyGotway committed:
Per #392, add -tcdiag and -lsdiag command line options for TC-Pairs and add vx_tc_util/diag_file.h/.cc. This compiles but the code doesn't actually do much yet.

JohnHalleyGotway committed:
Per #392, run bootstrap on seneca to add diag_file.h/.cc to vx_tc_util/Makefile.in.

JohnHalleyGotway committed:
Per #392, just including the new diag_file.h in vx_tc_util.h

JohnHalleyGotway committed:
Per #392, rename tc_diag.h/.cc to diag_file.h/.cc

JohnHalleyGotway committed:
Per #392, updating Makefile.in accordingly.

JohnHalleyGotway committed:
Per #392, fix capitalization bug.

JohnHalleyGotway committed:
Per #392, work in progress. Parsing the tcdiag and lsdiag files well. Next I need to read the diag data into the tracks.

JohnHalleyGotway committed:
Per #392, add logic to read lines of diagnostics data.

JohnHalleyGotway committed:
Per #392, update TrackInfo class to add the diagnostic data to the tracks. Still need to do more work there.

JohnHalleyGotway committed:
Per #392, switch ANV to GFS in the diagnostic file code.

JohnHalleyGotway committed:
Per #392, update TrackPoint class to store the actual diagnostic name and value. That makes the logic easier.

JohnHalleyGotway committed:
Per #392, remove the DiagName setting from TrackInfo class since we're storing it in TrackPoint instead.

JohnHalleyGotway committed:
Per #392, update TC-Pairs to add the diag data to the parsed tracks.

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_392_tcdiag_line_type

JohnHalleyGotway committed:
Per #392, restructure the logic for storing the TC diagnostics data by parsing it into an STL map first. If diag_name is an empty list, write all available diagnostics to the output.

JohnHalleyGotway committed:
Per #392, add TCDIAG header columns for tc_stat and the unit tests to use.

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_392_tcdiag_line_type

JohnHalleyGotway committed:
Per #392, still more work to do. But for TC-Stat, I'm now tracking the input TCMPR and TCDIAG lines separately.

JohnHalleyGotway committed:
Per #392, couple for changes for the previous commit.

JohnHalleyGotway committed:
Per #392, add support -diag_thresh and -init_diag_thresh job filtering options. Switch from reporting the number of lines read/kept to the number of track points to make the logic easier. This works for the filter job type, but still need to test it for other job types.

JohnHalleyGotway committed:
Per #392, update the tc_stat summary job to process TCDIAG values.

JohnHalleyGotway committed:
Per #392, rename TCLineCounts struct to TCPointCounts to clarify that we're now counting input track points, not lines.

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_392_tcdiag_line_type

JohnHalleyGotway committed:
Per #392, add more error checking. Each TCDIAG line must be preceded by a TCMPR. If they don't match, error out with a useful error message.

JohnHalleyGotway committed:
Per #392, add LineDataFile::peek_line() function and then update tc_stat_files.cc to call it when parsing .tcst files. Need to check for and parse trailing TCDIAG lines. ci-run-unit

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_392_tcdiag_line_type

JohnHalleyGotway committed:
Per #392, working on error messages.

JohnHalleyGotway committed:
Per #392, for TC-Stat add diag_thresh_name/val and init_diag_thresh_name/val options to the config file. For TC-Pairs, add -tcdiag and -lsdiag to the usage statement. For both, update the User's Guide.

JohnHalleyGotway committed:
Per #392, migrate new TC-Pairs and TC-Stat config options to other config files.

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_392_tcdiag_line_type

JohnHalleyGotway committed:
Per #392, fix parsing of the lsdiag input files.

JohnHalleyGotway committed:
Per #392, switch from using DiagMap to DiagName and DiagVal to maintain the order by which the values are read and stored.

JohnHalleyGotway committed:
Per #392, add tc_pairs.xml tests for diagnostics. Still need to refine them and add tc_stat tests.

JohnHalleyGotway committed:
Per #392, tweak log message.

JohnHalleyGotway committed:
Per #392, the lsdiag corresponds to ADECK SHIP tracks, as noted in https://ftp.nhc.noaa.gov/atcf/README

JohnHalleyGotway committed:
Per #392, add tcdiag_convert_map and lsdiag_convert_map entries to the TC-Pairs configuration file. Still need to figure out how to handle lsdiag values that require special handling.

JohnHalleyGotway committed:
Per #392, move the diagnostic location checking from TrackPoint up into TrackInfo and only check it once. This reduces the warning messages from 1500+ down to 7 in a single run. Only check locations for the first diagnostic processed.

JohnHalleyGotway committed:
Per #392, for -tcdiag and -lsdiag, let model=STRING be set to a comma-separated list of values.

JohnHalleyGotway committed:
Per #392, just updating code comments.

JohnHalleyGotway committed:
Per #392, print a single warning from tc_stat for each diagnostic name that is requested but is not actually present in the input data.

JohnHalleyGotway committed:
Per #392, refine the logic for tcdiag_convert_map config entry. Let it specify the conversion function based on the diagnostic name or units string. If both are provided, the conversion function for the name takes precedence.

JohnHalleyGotway committed:
Per #392, add logic to skip over lsdiag diagnostics that require special handling.

JohnHalleyGotway committed:
Per #392, update unit_tc_pairs.xml

JohnHalleyGotway committed:
Per #392, revert TCPairsConfig_ALAL2010 back to what it should be.

JohnHalleyGotway committed:
Per #392, update unit_tc_pairs.xml to attach lsdiag to OFCL and SHIP tracks.

JohnHalleyGotway committed:
Per #392, revise the DiagFile logic. Call clear() at the beginning of each read function. Improved handling of setting the Technique.

JohnHalleyGotway committed:
Per #392, add unit tests.

JohnHalleyGotway committed:
Per #392, adding tc_stat config file for it's unit test.

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_392_tcdiag_line_type

JohnHalleyGotway committed:
Per #392, add DiagType enumeration along with utility functions to convert between strings and enumerated values.

JohnHalleyGotway committed:
Per #392, enhance the DiagFile class to make use of the DiagType enumeration.

JohnHalleyGotway committed:
Per #392, large changes to TC-Pairs, replacing lsdiag_convert_map and tcdiag_convert_map with a single diag_convert_map.

JohnHalleyGotway committed:
Per #392, update unit_tc_pairs.xml to use the single -diag command line option instead of -lsdiag and -tcdiag.

JohnHalleyGotway committed:
Per #392, update TrackInfo to store the DiagType source.

JohnHalleyGotway committed:
Per #392, add SOURCE column to the TCDIAG line type from tc_pairs.

JohnHalleyGotway committed:
Per #392, update TC-Pairs documentation based on changes to the TCDIAG line type.

JohnHalleyGotway committed:
Per #392, when using a map of maps, the Docker build step errors out with:
tc_pairs_conf_info.h:118:60: error: '>>' should be '> >' within a nested template argument list
       std::map<DiagType,std::map<ConcatString,UserFunc_1Arg>> DiagConvertMap;
Adding the extra space to make the code compile in Docker.

JohnHalleyGotway committed:
Per #392, set the default lsdiag ATCFID as OFCL.

georgemccabe commented on internal/test_unit/xml/unit_tc_pairs.xml:
```diff
@@ -182,4 +182,27 @@
+      <pair><name>STORM_ID</name> <value>"AL092022"</value></pair>
+      <pair><name>INIT_BEG</name> <value>"20220926_00"</value></pair>
+      <pair><name>INIT_END</name> <value>"20220926_18"</value></pair>
+      <pair><name>DIAG_NAME</name> <value></value></pair>
```
It looks like there is no value set for DIAG_NAME that is used to set diag_name in the TCPairsConfig_DIAGNOSTICS file. It would be useful to see an example of what is a valid value for adding this functionality to the METplus wrapper. Is it a string surrounded by quotation marks? Adding a test where a value is set here would be useful to ensure that it works as expected.

JohnHalleyGotway replied:
@georgemccabe great point. When `diag_name` is empty, ALL diags from the input are passed to the output. I updated the test to provide a specific list with some names for the TCDIAG input and other names for the LSDIAG_RT input. That's a better test. I did manually test to confirm that an empty list still results in all being written, and it does.

Thanks for catching that.

JohnHalleyGotway replied:
FYI, you can ignore the ascii2nc diff. Those are from a recent PR that corrected an recent bug in ascii2nc. I just haven't updated develop-ref yet. Was hoping to include diffs from a second open PR in that update step.

georgemccabe commented on data/config/TCStatConfig_default:
```diff
@@ -134,6 +134,20 @@ init_str_val  = [];
+// storm diagnostic values in the TCDIAG lines.
+//
+diag_thresh_name = [];
+diag_thresh_val  = [];
```
What are valid values that can go here? Thresholds in numeric (e.g. gt) notation only? It would be useful to have an example in a unit test to add support in the METplus wrapper.

JohnHalleyGotway replied:
Please see this newly updated [unit test](https://github.com/dtcenter/MET/blob/bc2b5559fd08e8b41e70938a5c7f0e4540a27a98/internal/test_unit/xml/unit_tc_pairs.xml#L191) for an example. `diag_name` is an array of strings defining which diagnostics should be extracted from the the input files.

georgemccabe replied:
Sorry, I was asking about diag_thresh_val. Should this be a list of thresholds using only alphanumeric format? e.g. `[ gt4, ge5 ]`

JohnHalleyGotway replied:
Yes, as we discussed in the 10/27/2022 MET development meeting, here's an example:
```
diag_thresh_name = [ "TPW", "SHRG" ];
diag_thresh_val = [ gt50, le100 ];
```
Marking as resolved.

georgemccabe reviewed (changes requested):
I created a [METplus issue](https://github.com/dtcenter/METplus/issues/1898) to add support for the new options. I have a few questions that I added inline that would help implement these changes in the wrappers regarding the valid values that can be set for the new options.

I also suggested that unit tests that set some of these values would be useful not only for implementing the wrapper but ensuring that the values are read properly.

On first read of the documentation I thought that there were new command line arguments to tc_stat (-diag_thresh and -init_diag_thresh) but upon reread it sounds like those are job arguments. I'm not sure if it was initially unclear or if I was just skimming and missed that note.

I also see a TODO item on line 20 of [docs/Users_Guide/tc-pairs.rst](https://github.com/dtcenter/MET/pull/2315/files#diff-c82e4b8927f46f2738c4fd3a58080ffe738362e86f33708bef19da1aff00249d). I'm not sure if there is an issue to complete this or if it was missed, but we should make sure that we don't forget to finish that part.

In the GHA tests, I see that there is new tc_pairs and tc_stat output from the new test, which is expected. However, I also see differences in ascii2nc output. That doesn't seem related to these changes, but are they expected from another set of changes?

https://github.com/dtcenter/MET/actions/runs/3332482875/jobs/5513861007#step:5:119

```
COMPARING ascii2nc/airnow/HourlyData_20220312.nc
file1: /data/output/met_test_truth/ascii2nc/airnow/HourlyData_20220312.nc
file2: /data/output/met_test_output/ascii2nc/airnow/HourlyData_20220312.nc
ERROR: NetCDF headers differ:
6,7c6,7
< 	nhdr = 3638 ;
< 	nobs = 13737 ;
---
> 	nhdr = 3391 ;
> 	nobs = 13078 ;
9c9
< 	nhdr_sid = 3638 ;
---
> 	nhdr_sid = 3391 ;
```

JohnHalleyGotway committed:
Per #392, update diagnostics tests to specify the diag_name strings. Also needed to refine the logic in track_info.cc.

georgemccabe reviewed (approved):
Everything looks good. See my previous review for details.

jvigh commented on data/config/TCPairsConfig_default:
```diff
@@ -137,6 +137,34 @@ watch_warn = {
+     convert(x) = x / 10; },
+
+   { source = "LSDIAG_RT";
+     key = [ "CSST", "RSST", "DSST", "DSTA", "XDST", "XNST", "NSST", "NSTA",
```
Should LAT and LON be in this list? Are are these used for matching with the track object and not included? They are also divided by 10. I kind of think these should be included since you are including TLAT and TLON.

JohnHalleyGotway replied:
The LAT and LON lines are parsed as part of the "header" information. In fact, if they are missing from the input diag file, it should error out. As such, dividing LAT and LON by 10 is done automatically without any user option to do otherwise. I think that behavior makes sense.

JohnHalleyGotway replied:
@jvigh, OK, as requested I just pushed changes to let LAT and LON be included in the values written to the TCDIAG line. You can inspect those changes in [this commit](https://github.com/dtcenter/MET/commit/56f97c8aa2d459cec0ed5856fa85c75322e7b2c6).

Note that I did test manually that when I include `"LAT", "LON"` in `diag_name`, they do appear in the output TCDIAG line. And they also do appear when its set to an empty list to get *ALL* diagnostics, `diag_name = [];`.

Once you're satisfied with these changes, please actually add your "approval" on this PR. That adds a little green checkmark by your name is how you tell me that you're ready to proceed with these changes.

jvigh commented on data/config/TCPairsConfig_default:
```diff
@@ -137,6 +137,34 @@ watch_warn = {
+     key = [ "VVAV", "VMFX", "VVAC" ];
+     convert(x) = x / 100; }
+];
+
```
There should be another section like the following:

{ source = "LSDIAG_RT";
     key = [ "Z850", "D200","TGRD","DIVC" ];
     convert(x) = x / 10000000; }
];

jvigh replied:
{ source = "LSDIAG_RT";
key = [ "TADV" ];
convert(x) = x / 1000000; }
];

jvigh replied:
{ source = "LSDIAG_RT";
key = [ "PENC" ];
convert(x) = x / 10 + 1000.; }
];

JohnHalleyGotway replied:
Updated `TCPairsConfig_default` with these changes. Note that I included both "PENC" and "PENV" in the last conversion based on the docs:
```
{ source = "LSDIAG_RT";
key = [ "PENC", "PENV" ];
convert(x) = x / 10 + 1000; }
```

jvigh commented on data/config/TCPairsConfig_default:
```diff
@@ -137,6 +137,34 @@ watch_warn = {
+             "NTMX", "NTFR", "U200", "U20C", "V20C", "E000", "EPOS", "ENEG", "EPSS", "ENSS",
+             "T000", "TLAT", "TLON", "TWAC", "TWXC", "G150", "G200", "G250", "V000", "V850",
+             "V500", "V300", "PENC", "SHDC", "SHGC", "T150", "T200", "T250", "SHRD", "SHRS",
+             "SHRG", "PENV", "HE07", "HE05", "PW01", "PW02", "PW03", "PW04", "PW05", "PW06",
```
PENV should not be in this list, as it has a more complex units conversion. See note below.

JohnHalleyGotway replied:
Moved both "PENC" and "PENV" out of this last and down into their own conversion function entry.

jvigh commented on docs/Users_Guide/tc-pairs.rst:
```diff
@@ -235,6 +254,46 @@ ____________________
+
+The **diag_name** entry specifies a comma-separated list of strings for the tropical cyclone diagnostics of interest. This applies when the **-tcdiag** and/or **-lsdiag** command line options have been used to provide storm diagnostics data. If a non-zero list of diagnostic names is specified, only those diagnostics appearing in the list are written to the TCDIAG output line type. If defined as a ...
+
+A TCMPR line is written to the output for each track point. If diagnostics data is also defined for that track point, a TCDIAG line is written immediately after the corresponding TCMPR line. The contents of that TCDIAG line is deteremined by diagnostic names requrested in the **diag_name** entry.
```
deteremined -> determined
requrested -> requested

JohnHalleyGotway replied:
Fixed... thanks! Really wish spell checking was easier in this code!

jvigh commented on docs/Users_Guide/tc-pairs.rst:
```diff
@@ -235,6 +254,46 @@ ____________________
+
+.. code-block:: none
+
+  diag_convert_map = [
+     { source = "TCDIAG";
+       key = [ "(10C)", "(10KT)", "(10M/S)" ];
+       convert(x) = x / 10; },
+
+     { source = "LSDIAG_RT";
+       key = [ "CSST", "RSST", "DSST", "DSTA", "XDST", "XNST", "NSST", "NSTA",
+               "NTMX", "NTFR", "U200", "U20C", "V20C", "E000", "EPOS", "ENEG", "EPSS", "ENSS",
+               "T000", "TLAT", "TLON", "TWAC", "TWXC", "G150", "G200", "G250", "V000", "V850",
+               "V500", "V300", "PENC", "SHDC", "SHGC", "T150", "T200", "T250", "SHRD", "SHRS",
+               "SHRG", "PENV", "HE07", "HE05", "PW01", "PW02", "PW03", "PW04", "PW05", "PW06",
+               "PW07", "PW08", "PW09", "PW10", "PW11", "PW12", "PW13", "PW14", "PW15", "PW16",
+               "PW17", "PW18", "PW20", "PW21" ];
+       convert(x) = x / 10; },
+
+     { source = "LSDIAG_RT";
+       key = [ "VVAV", "VMFX", "VVAC" ];
+       convert(x) = x / 100; }
+  ];
```
This block in the documentation would need to be updated if the changes I suggested are made in the underlying config file.

JohnHalleyGotway replied:
Migrated changes from `TCPairsConfig_default` over into `tc-pairs.rst`. Thanks for the reminder.

jvigh commented on docs/Users_Guide/tc-pairs.rst:
```diff
@@ -235,6 +254,46 @@ ____________________
+
+The **diag_convert_map** entries define conversion functions to be applied to diagnostics data read with the **-diag** command line option. Each array element is a dictionary consisting of a **source**, **key**, and **convert(x)** entry.
+
+The **source** is one of the supported diagnostics data sources. The **key** is an array of strings. The strings can specify diagnostic names or units, although units are only checked for **TCDIAG** sources. If both the name and units are specified, the conversion function for the name takes precedence. **convert(x)** is a function of one variable which defines how the diagnostic data should be c ...
```
who -> whose

JohnHalleyGotway replied:
Done.

jvigh commented on docs/Users_Guide/tc-stat.rst:
```diff
@@ -281,7 +281,25 @@ _________________________
+  diag_thresh_name = [];
+  diag_thresh_val  = [];
+
+The **diag_thresh_name** and **diag_thresh_val** fields stratify individual track points by applying thresholds to numeric data columns from the TCDIAG lines. Specify a comma-separated list of diagnostics names and thresholds to be applied. The length of **diag_thresh_val** should match that of **diag_thresh_name**. If the storm diagnostic does not meet the threshold, discard both the TCMPR and T ...
```
Is it a given that this logic also applies to non-track points (e.g., intensity or other diagnostic values)? 

To say it another way: The diagnostic values are attached to the tracks, so then does it follow that any diagnostic values also get stratified by these criteria?

JohnHalleyGotway replied:
Yes, that is true. In fact, when TC-Stat reads TCMPR and TCDIAG lines, it combines them into the same TrackPoint/TrackInfo class. So when I say "track points" I'm implying track location/intensity/diagnostic information.

But if you can think of a better way to state that more clearly, I'm happy to change it however you'd like.

jvigh commented on internal/test_unit/config/TCPairsConfig_ALAL2010:
```diff
@@ -128,6 +128,11 @@ watch_warn = {
    time_offset = -14400;
 }
 
```
Is there a reason the stormid in this file is "ALAL2010"? Normally, it'd just be "AL2010". The same question applies to the name given to this file.

JohnHalleyGotway replied:
So the STORM_ID is defined as "BBCCYYYY" where BB is the 2-letter basin, CC is the 2-digit storm, and YYYY is the 4 digit year. This [comment](https://github.com/dtcenter/MET/blob/82e5833f443e7e8c1cffe9071887b7a8cb5e3c4c/src/libcode/vx_tc_util/track_info.cc#L1075) followed by this [line of code](https://github.com/dtcenter/MET/blob/82e5833f443e7e8c1cffe9071887b7a8cb5e3c4c/src/libcode/vx_tc_util/track_info.cc#L1097) explains the behavior. We tread CC = AL (i.e. "ALL") in the STORM_ID as basically being a wildcard, but I don't see this documented in the User's Guide.

I don't know how we ended up with that convention. Not sure if we inherited from the original NHC vx code or if I just made it up along the way? We could ask @KathrynNewman, but I'd recommend addressing it outside of this pull request.

KathrynNewman replied:
I think this way of representing all storms (rather than a single storm) was legacy from the NHC verification system, but I'm not 100% - I'll see if I can dig this out and follow-up (outside this PR).

jvigh commented on internal/test_unit/config/TCStatConfig_ALAL2010:
```diff
@@ -136,6 +136,20 @@ init_str_val  = [];
 init_str_exc_name = [];
 init_str_exc_val  = [];
 
```
Similarly, any reason why the STORMID is ALAL2010?

JohnHalleyGotway replied:
See previous explanation.

jvigh commented on src/libcode/vx_tc_util/diag_file.h:
```diff
@@ -0,0 +1,144 @@
+
+////////////////////////////////////////////////////////////////////////
+//
+// LSDIAG files:
```
Real-time LSDIAG files

JohnHalleyGotway replied:
Done. Thanks.

jvigh reviewed (changes requested):
@JohnHalleyGotway  - I've looked through all the changes. This is an impressive amount of development. My main comments were in the config where the unit conversions are defined. Some variables were missed. I also added a few suggestions/corrections for the documentation. Overall, it looks good!

JohnHalleyGotway committed:
Per #392, updates based on PR #2315 review from @jvigh. Thanks for the careful review Jonathan.

JohnHalleyGotway commented:
@jvigh thanks for the detailed review! I have addressed all the issues you raised, and clicked the button to "re-request" your review. Please "approve" this PR once you've confirmed that my updates are what you intended.

JohnHalleyGotway committed:
Per #392, hopefully the final commit for this feature branch. These changes enable LAT and LON values to be included on the TCDIAG line as requested in Jonathan's PR review. Note that the longitude values stored for the header are converted from degrees east to west (*-1) and rescaled to a standard range to match the track data. The LON values written to the TCDIAG line are NOT converted in this way. Figure its better to leave them as close to the input as possible. Also renamed the input test data directory from lsdiag to lsdiag_rt, figuing that we'll need add data for lsdiag_dev eventually.

KathrynNewman reviewed (approved):
Testing on seneca of tc_pairs and tc_stat works (and fails) as expected. I noted that adding a bogus DIAG_NAME does nothing (no warning or output) - but after discussion with @JohnHalleyGotway, this is reasonable behavior. Calling LSDIAG_DEV resulted in errors and no matches as expected (additional development is needed for this source). Overall, looks good! Once we have more clarity on the tracks, I'll want to sanity check that logic closer.
